janus-idp/operator
# fix: Fix reconciliation error preventing existing Backstage operands from being upgraded when the operator is upgraded from `1.1.x` [RHIDP-2597]

## Description

This PR fixes a reconciliation error occurring against existing CRs after upgrading the operator from the released `1.1.x` branch to either `main` or `1.2.x`.

This issue is currently due to a PATCH operation that was denied by Kubernetes against the existing database Service object.
https://github.com/janus-idp/operator/pull/369 (fix for https://issues.redhat.com/browse/RHIDP-2432) introduced a mechanism to "replace" `StatefulSet` objects when the PATCH operation fails, by deleting it (without deleting its dependents), so that the next reconciliation will recreate it.
To prevent similar issues (patching fields that are not allowed by K8s) from occurring again in the future, this PR generalizes this logic ~based on the patch error status and cause (either Field Update Forbidden or Field Value Invalid)~, i.e., it tries to delete the object anyway if patching it failed, so that it can be recreated upon the next reconciliation.

## Which issue(s) does this PR fix or relate to

- Fixes #382 
- Fixes https://issues.redhat.com/browse/RHIDP-2597

## PR acceptance criteria


## How to test changes / Special notes to the reviewer
Repro steps provided in #382.

- To reproduce the issue using the E2E tests just run `make test-e2e-upgrade`. It will leverage the default operator image (not patched) and will not pass.
- Run `make test-e2e-upgrade IMG=quay.io/janus-idp/operator:0.3.0-pr-384` to see the E2E tests passing using the operator image from this PR.
